_posts/ann_catton/2019-03-25-introspection.md

```diff
@@ -0,0 +1,155 @@
+
+To get started, we used the meta-field `__schema` ([reserved](https://github.com/facebook/graphql/blob/master/spec/Section%204%20--%20Introspection.md#reserved-names){:target="\_blank"} names in the introspection system are prefixed with `__`), which returns a list of all the types in your schema. This gives you a good overview of which types you can query:
+
+```javascript
+{
+  __schema {
+    types {
+      name
+    }
+  }
+}
+```
+
+And the (abbreviated) response:
+
+```javascript
+{
+  "data": {
+    "__schema": {
+      "types": [
+        {
+          "name": "Root"
+        },
+        {
+          "name": "ID"
+        },
+        {
+          "name": "Node"
+        },
+        {
+          "name": "User"
+        },
+        {
+          "name": "String"
+        },
+        {
+          "name": "QueryResults"
+        },
+        {
+          "name": "RepositoryCases"
+        },
+        ...
+      }
+    }
+```
+
+Once you have a list of types, you can then use the meta-field `__type` to get more information about a specific type by using a Type Name introspection query:
+
+```javascript
+{
+  __type(name: String!) {
+    name
+    fields {
+      name
+      description
+    }
+  }
+}
+```
+
+Depending on your schema, you may need to add more levels of nesting to your query. The information we were trying to get at is another level down, so we had to repeat ourselves a little. We first tried querying type `"Case"`, as all clinical fields are nested under this type in our [case centric index](https://github.com/NCI-GDC/gdc-models/blob/master/es-models/case_centric/case_centric.mapping.yaml){:target="\_blank"}:
+
+```javascript
+{
+  __type(name: "Case") {
+    fields {
+      name
+      description
+      type {
+        name
+        fields {
+          name
+          description
+          type {
+            name
+          }
+        }
+      }
+    }
+  }
+}
+```
+
+However, the result didn’t go very far down the tree; what was under `"Case"` was mostly metadata fields, and we didn’t find a lot of the fields we were looking for from the models. We dug around some more, trying instead more specific types from our clinical data, such as `"Treatment"` and `"Diagnosis"`. I thought we'd hit the jackpot with this one; it returned a lot of data that _looked_ like what we needed - a good list of fields with their descriptions and types, without a lot of metadata fields we didn’t need. But comparing the results to our models, we soon discovered this was coming from our [data dictionary](https://github.com/NCI-GDC/gdcdictionary/tree/develop/gdcdictionary/schemas){:target="\_blank"}, and had more fields than what was in the index. We’d end up with a different version of the same issue. We needed to query a type that had its source in our case index.
```

Comment:
"_...without a lot of metadata fields we didn't need._"

Sorry, I know @francisgerthoffert made you change this already, but the grammar confused me a bit here, I had to re-read it a couple times for it to click.  I think "...without a lot of..." was a bit of a mouthful.

Not a big deal and it's quite subjective, but maybe an alternative like, "_...but without metadata fields that we didn't need._"

P.S. I liked the original "ruder" version better :P.

Reply:
Hehe. I like your version, it's cleaner!